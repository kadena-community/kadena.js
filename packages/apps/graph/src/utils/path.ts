import {
  mkdirSync,
  readdirSync,
  renameSync,
  rmdirSync,
  statSync,
  unlinkSync,
  writeFileSync,
} from 'fs';
import { basename, join } from 'path';

export async function createDirAndWriteFile(
  dir: string,
  fileName: string,
  data: string,
): Promise<void> {
  mkdirSync(dir, { recursive: true });
  writeFileSync(join(dir, fileName), data);
}

export async function clearDir(dir: string, extension?: string): Promise<void> {
  const files = readdirSync(dir);
  for (const file of files) {
    const filePath = join(dir, file);
    try {
      if (statSync(filePath).isDirectory()) {
        await clearDir(filePath, extension); // Recursive call
        rmdirSync(filePath); // Remove the directory itself
      } else if (!extension || file.endsWith(extension)) {
        unlinkSync(filePath);
      }
    } catch (error) {
      console.error(`Error processing file ${filePath}: ${error}`);
    }
  }
}
export async function flattenFolder(
  basePath: string,
  fileExtensions: string[],
  currentPath: string = basePath,
): Promise<void> {
  // Read directory
  const files = readdirSync(currentPath);
  for (const file of files) {
    const filePath = join(currentPath, file);

    // if directory, recurse and delete folder
    if (statSync(filePath).isDirectory()) {
      await flattenFolder(basePath, fileExtensions, filePath);
      rmdirSync(filePath);
    } else if (fileExtensions.some((ext) => file.endsWith(ext))) {
      // if file, rename to include parent folder name
      if (basePath === currentPath) continue;
      const relativePath = currentPath.replace(basePath, '');
      const folderChain = relativePath.split('/').filter(Boolean).join('.');
      const newFilePath = join(
        basePath,
        `${basename(folderChain)}.${basename(file)}`,
      );
      renameSync(filePath, newFilePath);
    }
  }
}
