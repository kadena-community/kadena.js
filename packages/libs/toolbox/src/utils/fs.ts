import { mkdir, stat, writeFile as writeFileN } from 'node:fs/promises';
import { dirname } from 'node:path';

export const writeFileAtPath: typeof writeFileN = async (
  path,
  data,
  options,
) => {
  if (typeof path !== 'string') throw new Error('path must be a string');
  const exist = await stat(path)
    .then(() => true)
    .catch(() => false);
  if (!exist) {
    const dir = dirname(path);
    await mkdir(dir, { recursive: true });
  }
  return writeFileN(path, data, options);
};
