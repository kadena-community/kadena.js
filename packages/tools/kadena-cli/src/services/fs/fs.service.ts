import type { Dirent, Stats } from 'fs';
import fs from 'node:fs/promises';

export interface IFileSystemService {
  readFile: (path: string) => Promise<string | null>;
  writeFile: (path: string, data: string) => Promise<void>;
  deleteFile: (path: string) => Promise<void>;
  deleteDirectory: (path: string) => Promise<void>;
  directoryExists: (path: string) => Promise<boolean>;
  fileExists: (path: string) => Promise<boolean>;
  ensureDirectoryExists: (path: string) => Promise<void>;
  readDir: (path: string) => Promise<string[]>;
  readDirWithTypes: (path: string) => Promise<Dirent[]>;
  appendFile: (path: string, data: string) => Promise<void>;
  lstat: (path: string) => Promise<Stats>;
}

export const fileSystemService: IFileSystemService = {
  async readFile(path: string) {
    try {
      return await fs.readFile(path, 'utf8');
    } catch (e) {
      return null;
    }
  },
  async writeFile(path: string, data: string) {
    await fs.writeFile(path, data, 'utf8');
  },
  async deleteFile(path: string) {
    await fs.unlink(path);
  },
  async deleteDirectory(path: string) {
    await fs.rm(path, { recursive: true });
  },
  async directoryExists(path: string) {
    try {
      const stat = await fs.stat(path);
      return stat.isDirectory();
    } catch {
      return false;
    }
  },
  async fileExists(path: string) {
    try {
      const stat = await fs.stat(path);
      return stat.isFile();
    } catch {
      return false;
    }
  },
  async ensureDirectoryExists(directory: string) {
    if (await fileSystemService.directoryExists(directory)) return;

    await fs.mkdir(directory, {
      recursive: true,
    });
  },
  async readDir(path: string) {
    return fs.readdir(path);
  },
  async readDirWithTypes(path: string) {
    return fs.readdir(path, { withFileTypes: true });
  },
  async appendFile(path: string, data: string) {
    await fs.appendFile(path, data, { encoding: 'utf8' });
  },
  async lstat(path: string) {
    return fs.lstat(path);
  },
};
