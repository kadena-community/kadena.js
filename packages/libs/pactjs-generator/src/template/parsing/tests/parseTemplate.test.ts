import { parseTemplate } from '../parseTemplate';

import { readFileSync } from 'fs';
import { join } from 'path';

describe('parseTemplate', () => {
  it('parses a simple template `Hello {{name}}`', () => {
    const result = parseTemplate(`Hello {{name}}`);
    const expected = { parts: ['Hello ', ''], holes: ['name'] };
    expect(result).toEqual(expected);
  });

  it('parses a command template', () => {
    const result = parseTemplate(
      `(coin.transfer "{{sender}}", "{{receiver}}", {{amount}})`,
    );
    const expected = {
      parts: ['(coin.transfer "', '", "', '", ', ')'],
      holes: ['sender', 'receiver', 'amount'],
    };
    expect(result).toEqual(expected);
  });

  it('parses a JSON string template', () => {
    const template = readFileSync(
      join(__dirname, './simple-transfer.yaml'),
      'utf8',
    );
    const result = parseTemplate(template);
    const expected = {
      parts: [
        `code: |-
  (coin.transfer \"`,
        `\" \"`,
        `\" `,
        `)
data:
publicMeta:
  chainId: \"`,
        `\"
  sender: `,
        `
  gasLimit: 2500
  gasPrice: 0.00000001
  ttl: 7200
networkId: `,
        `
signers:
  - public: `,
        `
    caps:
      - name: \"coin.TRANSFER\"
        args: [`,
        `, `,
        `, `,
        `]
      - name: \"coin.GAS\"
        args: []
type: exec
docs:
  from-acct: Account that you want to send from
  to-acct: Account that you want to send to
  amount: The amount of KDA that you want to transfer
  chain: The chain that you want to execute the transaction on
  network: The network that you want to use (\"testnet04\", \"mainnet\")
  from-key: The key (without k:) that signs the caps
`,
      ],
      holes: [
        'from-acct',
        'to-acct',
        'amount',
        'chain',
        'from-acct',
        'network',
        'from-key',
        'from-acct',
        'to-acct',
        'amount',
      ],
    };
    expect(result).toEqual(expected);
  });
});
