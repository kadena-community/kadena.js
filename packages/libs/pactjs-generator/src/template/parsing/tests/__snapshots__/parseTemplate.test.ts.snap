// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTemplate parses a YAML template 1`] = `
Object {
  "holes": Array [
    "from-acct",
    "to-acct",
    "amount",
    "chain",
    "from-acct",
    "network",
    "from-key",
    "from-acct",
    "to-acct",
    "amount",
  ],
  "parts": Array [
    "code: |-
  (coin.transfer \\"",
    "\\" \\"",
    "\\" ",
    ")
data:
publicMeta:
  chainId: \\"",
    "\\"
  sender: ",
    "
  gasLimit: 2500
  gasPrice: 0.00000001
  ttl: 7200
networkId: ",
    "
signers:
  - public: ",
    "
    caps:
      - name: \\"coin.TRANSFER\\"
        args: [",
    ", ",
    ", ",
    "]
      - name: \\"coin.GAS\\"
        args: []
type: exec
docs:
  from-acct: Account that you want to send from
  to-acct: Account that you want to send to
  amount: The amount of KDA that you want to transfer
  chain: The chain that you want to execute the transaction on
  network: The network that you want to use (\\"testnet04\\", \\"mainnet\\")
  from-key: The key (without k:) that signs the caps
",
  ],
}
`;
